### PR TITLE
Add function library to FieldArray, and use FieldArray for loading samples from inference

### PR DESCRIPTION
**Pull request by cdcapano**
This patch does two major things: first, it adds a function library attribute to `FieldArray`, so that functions beyond what's in numpy may be used. Right now, the library includes all of the functions in `conversions.py`, `coordinates.py`, and `cosmolgy.py` that are specified in each module's `__all__` list. Those functions can now be called as items; e.g., if `arr` is a `FieldArray` or `WaveformArray` that has fields `mass1` and `mass2`, you can do `arr['eta_from_mass1_mass2(mass1, mass2)']` to get `eta`.

Second, this patch fixes an issue with the inference plotting codes in which you could not easily plot mass1 and mass2 if you sampled in chirp mass and mass ratio. The big problem is that `WaveformArray` assumes it has fields called `mass1` and `mass2`, from which it derives `mchirp` and `q`, not the other way around. I decided the best way to deal with this is to have the samples loaded as a `FieldArray` instead of a `WaveformArray`. This means that the sample array will not have pre-defined virtual fields; instead, you need to use the functions in the function library to derive parameters. This sort of makes sense; we are not always doing PE with CBC waveforms, and so there's no reason to expect any particular set of parameters to exist in a result file.

@cmbiwer This does mean you have to be a bit more explicit on the command line when using `plot_posterior`. For example, say you sampled in mass1 and mass2 but wanted to plot mchirp and eta. Before you could just do `--parameters mchirp eta`. Now you need to do `--parameters 'mchirp_from_mass1_mass2(mass1, mass2):$\mathcal{M}$' 'eta_from_mass1_mass2(mass1, mass2):$\eta$'`. Conversely, if you sampled in mchirp, q, but wanted to plot mass1, mass2, it'd be `--parameters 'mass1_from_mchirp_q(mchirp, q):$m_1$' 'mass2_from_mchirp_q(mchirp, q):$m_2$'` It's a bit more to type, but I think the clarity is worth it.

@ahnitz You and I had discussed removing the default virtual fields from WaveformArray entirely. However, since I'm using `FieldArray` for the inference results now, I decided to leave `WaveformArray`'s virtual fields alone. They are handy for loading a template bank, where the fields always are mass1, mass2, etc. This means that that place where you sort the template bank on chirp mass doesn't need to change.

This patch also fixes a bug I found in `FieldArray.__array_finalize__`. If you create a new array from a slice of a current array, that method should copy all of the instance settings from the old array to the new one. Instead, it was copying settings from the new array to the old, which would replace any attribute changes someone might have made to an instance (like adding a new virtual field) with the class's default attributes.

Finally, this patch renames `m_p` and `m_s` in `parameters.py` to `primary_mass` and `secondary_mass`, to be consistent with `conversions.py`.

### CODE REVIEW

File: pycbc/waveform/parameters.py [hunk 1]
```diff
@@ -205,39 +205,39 @@ def docstr(self, prefix='', include_label=True):
 q = Parameter("q",
                 dtype=float, label=r"$q$",
                 description="The mass ratio, m1/m2, where m1 >= m2.")
-m_p = Parameter("m_p",
-                dtype=float, label=r"$m_{\mathrm{pr}}$",
+primary_mass = Parameter("primary_mass",
```
Comment (ahnitz) [inline pycbc/waveform/parameters.py]: Why the change in notation from m_p to m_1?

Comment (cdcapano) [inline pycbc/waveform/parameters.py]: @duncan-brown had requested that we change the labelling to be `m_1/2` to be consistent with plots produced by lalinference. Having `m_{pr}` had confused Stuart; he thought it meant source mass, which it didn't.

Comment (ahnitz) [inline pycbc/waveform/parameters.py]: I assume it still means primary mass though, and *not* mass1 which could be the lighter one? That's my confusion, I guess.

Comment (ahnitz) [inline pycbc/waveform/parameters.py]: Anyway, it only affects the labeling and not the name, so I'll approve.

Comment (cdcapano) [inline pycbc/waveform/parameters.py]: @ahnitz Yeah, it still means the primary mass. I agree, it's a bit confusing, but that seems to be the convention that's been adopted by the collaboration. As you say, it only affects the labeling though, so can easily be overridden.

**Comment by cdcapano** [comment]
@cmbiwer Since I haven't heard any complaints from you, I'm going ahead and merging.
